File: isValid.py
# Time Complexity : O(n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
class Solution(object):
    def isValid(self, s):
        if len(s)%2 != 0:
            return False
        stack = []
        for i in s:
            if i == '(' or i == '{' or i == '[':
                stack.append(i)
            else:
                if stack:
                    if i == ')':
                        if stack.pop() != '(':
                            return False
                    elif i == ']':
                        if stack.pop() != '[':
                            return False
                    elif i == '}':
                        if stack.pop() != '{':
                            return False
                else:
                    return False
        if stack:
            return False

        return True

File: is-valid.java
// Time, Space - O(N), O(N)
class Solution {
    public boolean isValid(String s) {
        if(s == null || s.length()==0) {
            return true;
        }
        Stack<Character> st = new Stack<>();
        HashMap<Character, Character> map = new HashMap<>();
        map.put('(',')');
        map.put('{','}');
        map.put('[',']');
        for(int i=0;i<s.length();i++) {
            char c = s.charAt(i);
            if(c == '(' || c == '{' || c == '[') {
                st.push(map.get(c));
            }
            else if(st.isEmpty()||st.pop()!=c){
                return false;
            }
        }
        
        return st.isEmpty();
    }
}
